docs/export-keyword.md

```diff
@@ -0,0 +1,126 @@
+
+```luau
+local Library = require("Library")
+local point : Library.Point = { x = 0, y = 0 }
+```
+
+Right now, this is the only use of the export keyword but it would be great if we let users export more with it to offer more utility and make exporting library methods and values easier.
+
+## Design
+
+### Syntax
+You can prefix any non-local declaration with the export keyword in the top level scope of a module. For values this looks like:
+
+```luau
+export version = "5.1"
+```
+
+While methods use:
+
+```luau
+export function init()
+  -- do a thing
+end
+```
+
+### Behavior
+
+When a value or method is prefixed with export, it is automatically added to a hidden export table which is frozen as soon as the module returns.
+
+Take the following module:
+
+```luau
+export type Point = { x: number, y: number }
+
+export function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+```
+
+This essentially becomes sugar for:
+
+```luau
+local _EXT = {}
+
+type Point = { x: number, y: number }
+type _EXT.Point = Point -- note: this doesn't actually work today but one day it should
+
+local function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+_EXT.distance = distance
+
+return table.freeze(_EXT)
+```
+
+If the user attempts to assign to an exported identifier then we would throw an error explaining that the interface cannot be changed once it has been exported.
+
+### Nuances
+Due to the implementation, most things should "just-work". Here are some examples to consider:
+
+#### Calling an Exported Function
+You can call an exported function as it's registered as a local before being added to the export table.
+
+```luau
+export function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+
+distance({0, 0}, {1, 1})
+```
+
+#### Nested Tables
+You can export tables with additional values inside of them.
+
+```luau
+export triangle = {}
+
+function triangle.draw()
+
+end
+```
+
+Something important to note here is that the nested table, `triangle` is not frozen.
+
+#### Multiple Returns
+Today, you can use the export keyword along with a return statement at the end of your module. If you use the export keyword with a value however we will throw an error if you also attempt to return.
+
+```luau
+export function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+
+return table.freeze {
+  distance = distance
+}
+```
+
+## Drawbacks
+The keyword already exists for types so there's not much cost in us adding support to values. The main drawback is that it's another way to do module exports. We already have a way to do that, do we really need another?
+
+## Alternatives
+
+### Do Nothing
+It's already possible to export values from modules using a return statement. We don't actually need to do this, it's more of a nice-to-have.
```

Comment:
Note that it's not entirely syntactic sugar.

```lua
local mod = {}

function mod.distance(p1: Point, p2: Point)
  local x, y = a.X - b.X, a.Y - b.Y
  return math.sqrt(x * x + y * y)
end
```

If you wanted to call `mod.distance` somewhere else in this module, it will not get inlined because in theory, `mod.distance` could have been transmuted to a different function with a different implementation.

Reply:
This is true, and also worth calling out. I think it's unlikely we'd expose `_EXT` to users unless we think there's a valid use-case (maybe metatable stuff?)

Reply:
Yeah I don't want to expose `_EXT`. It's trivial for the VM to not expose it too, it already does this with loops which introduces 3 locals you can't even use.

Reply:
I also question the value of adding metatable to the export table. I've never seen anyone want to do this, and if they do they can use `return` with no `export`s.

Reply:
Agree with this so let's keep it locked down and if legitimate use-cases come up we can discuss as part of another RFC.